azurerm/internal/services/eventgrid/eventgrid_event_subscription_resource.go

```diff
@@ -254,12 +283,20 @@ func resourceEventGridEventSubscriptionRead(d *schema.ResourceData, meta interfa
 
 		d.Set("event_delivery_schema", string(props.EventDeliverySchema))
 
-		if azureFunctionEndpoint, ok := props.Destination.AsAzureFunctionEventSubscriptionDestination(); ok {
+		destination := props.Destination
+		if props.DeliveryWithResourceIdentity != nil {
+			destination = props.DeliveryWithResourceIdentity.Destination
+			if err := d.Set("delivery_identity", flattenIdentity(props.DeliveryWithResourceIdentity.Identity)); err != nil {
+				return fmt.Errorf("setting `delivery_identity` for EventGrid Event Subscription %q (Scope %q): %s", id.Name, id.Scope, err)
+			}
```

Comment:
since this could be referenced - rather than setting this inline, we'll need to ensure this is set in all cases - as such can we assign the flattened value to a variable then set that into the state outside of the block? (I'll push a commit for this)